### PR TITLE
check that amis exist

### PR DESCRIPTION
**Pull request by jhford**
@gregarndt might like this... we should check that given AMIs exist before creating the worker type!

### CODE REVIEW

File: lib/check-for-ami.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  if (result.Images.length !== 1) {
+    throw new Error('found more images than expected');
+  }
+
+  if (result.Images[0].ImageId === ami) {
+    return true;
+  } else {
+    throw new Error('something wacky happened');
```
Comment (gregarndt) [inline lib/check-for-ami.js]: Love the humor here, but if looking at logs, we might want to include something about what "wacky" is :)

File: lib/check-for-ami.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      assert(!region, 'this ec2 module doesnt understand region');
+      result = await ec2.describeImages(request);
+    }
+    debug('result from ec2: %j', result);
+  } catch (err) {
+    debug('did not find %s', ami);
+    return false;
+  }
+
+  if (result.Images.length !== 1) {
```
Comment (gregarndt) [inline lib/check-for-ami.js]: so, when searching for a particular ami, more than one could return?

File: lib/routes/v1.js [hunk 3]
```diff
@@ -169,6 +183,18 @@ api.declare({
 
   let modDate = new Date();
 
+  // We want to make sure that all AMIs that we are submitting are valid
+  for (let region of input.regions) {
+    let validAmi = await amiExists(this.ec2, region.launchSpec.ImageId, region.region);
+    if (!validAmi) {
```
Comment (gregarndt) [inline lib/routes/v1.js]: I'm not aware of the behavior, but what happens when the underlying code throws an error? does our api handling return back a certain status code then with the error?  Do we want to catch the error and then include the ami and region that was used in the response?

Comment (jonasfj) [inline lib/routes/v1.js]: API wrapper assumes handlers returns a promise... and if the promise is rejected, it'll return 500 with an incidentid, error message (which may be sensitive) is printed to log along with the incidentid.

File: lib/check-for-ami.js (hunk 1 above)
Comment (jhford) [inline lib/check-for-ami.js]: yah... true.

File: lib/check-for-ami.js (hunk 2 above)
Comment (jhford) [inline lib/check-for-ami.js]: With the ec2 api, i'm not going to rule out anything.  This is purely defensive programming.

**Comment by gregarndt** [comment]
:+1:

**Comment by jhford** [comment]
@jonasfj this might have bitrotted a bit, but what do you think about this check?

File: lib/check-for-ami.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      'all',
+    ],*/
+    ImageIds: [ami],
+  };
+
+  let result;
+  try {
+    // If we're using the multi-region wrapper, we handle things
+    // slightly differently
+    if (ec2.describeImages && ec2.describeImages.inRegion) {
```
Comment (jonasfj) [inline lib/check-for-ami.js]: Do we have a need for both? can't we just assert that you always give `region` and an instance of `multi-region wrapper` ?

File: lib/check-for-ami.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      result = await ec2.describeImages.inRegion(region, request);
+    } else {
+      assert(!region, 'this ec2 module doesnt understand region');
+      result = await ec2.describeImages(request);
+    }
+    debug('result from ec2: %j', result);
+  } catch (err) {
+    debug('did not find %s', ami);
+    return false;
```
Comment (jonasfj) [inline lib/check-for-ami.js]: Does it return an empty list if there is no matches or does it throw an error.
If it returns an empty list we should not handle the error, as it indicate inability to check for existence, right?

File: lib/routes/v1.js [hunk 6]
```diff
@@ -169,6 +183,18 @@ api.declare({
 
   let modDate = new Date();
 
+  // We want to make sure that all AMIs that we are submitting are valid
+  for (let region of input.regions) {
+    let validAmi = await amiExists(this.ec2, region.launchSpec.ImageId, region.region);
```
Comment (jonasfj) [inline lib/routes/v1.js]: I would do:
```js
let missing = [];
await Promise.all(input.regions.map(async (def) => {
  if (await amiExists(this.ec2, def.launchSpec.ImageId, def.region)) {
    missing.push({imageId: def.launchSpec.ImageId, region: def.region});
  }
}));
if (missing.length > 0) {
   .... // Send error
}
```
or even shorter:
```js
let missingAMI = _.every(await Promise.all(input.regions.map(def => {
  return amiExists(this.ec2, def.launchSpec.ImageId, def.region);
})));
if (missingAMI) {
   .... // Send error without saying which AMI is missing...
}
```

Either way, using `Promise.all` makes sense as we get the requests in parallel.. As the requests goes out to different regions, there is no reason not to make them in parallel.

**Comment by jonasfj** [comment]
Minor suggestions, generally I think it's nice to do this..
But we still have to survive cases where someone deletes an AMI after creating the workerType.

**Comment by jonasfj** [comment]
`r+`, but please do consider some of the suggestions...

File: lib/check-for-ami.js (hunk 4 above)
Comment (jhford) [inline lib/check-for-ami.js]:
```javascript
> a = {}
{}
> if (a.b.c) {
... console.log('hi');
... }
TypeError: Cannot read property 'c' of undefined
    at repl:1:8
    at REPLServer.defaultEval (repl.js:132:27)
    at bound (domain.js:254:14)
    at REPLServer.runBound [as eval] (domain.js:267:12)
    at REPLServer.<anonymous> (repl.js:279:12)
    at REPLServer.emit (events.js:107:17)
    at REPLServer.Interface._onLine (readline.js:214:10)
    at REPLServer.Interface._line (readline.js:553:8)
    at REPLServer.Interface._ttyWrite (readline.js:830:14)
    at ReadStream.onkeypress (readline.js:109:10)
```

I just like do be extra careful when checking.

File: lib/check-for-ami.js (hunk 5 above)
Comment (jhford) [inline lib/check-for-ami.js]: It throws an error, sadly.  There might be some property on the error object that will tell us this, but the outcome for both failing to check and not having the image is the same: we don't process the request.

File: lib/routes/v1.js (hunk 6 above)
Comment (jhford) [inline lib/routes/v1.js]: This was written when we were thinking that Promise.all might be related to provisioner freezing and I was avoiding using it.  Yes, we can do it with Promise.all!

**Comment by jhford** [comment]
The problem we currently don't have a way to bubble up errors from the provisioner background process to be visible to a consumer without dropping down to Heroku logs.

**Comment by jhford** [comment]
This should still be done, but someone will need to unbitrot this patch first.  I'll see if I can do it this week.

**Comment by jhford** [comment]
@jonasfj did you want another look?  if not, I'll be pushing this tomorrow

File: src/check-for-ami.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    ],*/
+    ImageIds: [ami],
+  };
+
+  let result;
+  try {
+    // If we're using the multi-region wrapper, we handle things
+    // slightly differently
+    if (ec2.describeImages && ec2.describeImages.inRegion) {
+      result = await ec2.describeImages.inRegion(region, request);
```
Comment (jonasfj) [inline src/check-for-ami.js]: didn't we drop multi-region-aws-thingy?

Comment (jonasfj) [inline src/check-for-ami.js]: Ì think you forgot to update this patch...

File: src/api-v1.js [hunk 8]
```diff
@@ -299,6 +317,20 @@ api.declare({
...
 
+  // We want to make sure that all AMIs that we are submitting are valid
+  let missing = [];
+  await Promise.all(input.regions.map(async (def) => {
+    if (await amiExists(this.ec2, def.launchSpec.ImageId, def.region)) {
+      missing.push({imageId: def.launchSpec.ImageId, region: def.region});
+    }
+  }));
+  if (missing.length > 0) {
+    return res.status(400).json({
```
Comment (jonasfj) [inline src/api-v1.js]: nit, if you are on a recent tc-lib-api, there is a reportError function which will do great here...

**Comment by jonasfj** [comment]
@jhford, I think there is some bitrot here... just tiny things..

File: src/check-for-ami.js (hunk 7 above)
Comment (jhford) [inline src/check-for-ami.js]: i wrote this patch knowing the drop would happen, but I'll remove the transitional code any way!

**Pull request merged**